In [106]:
import pandas as pd

df1 = pd.read_csv("../../../etl/raw_data/한국고전종합DB_관계망/itkc_events.csv")
df2 = pd.read_csv("../../../etl/raw_data/한국고전종합DB_관계망/itkc_event_relations.csv")

In [107]:
df1.shape

(1542, 9)

In [108]:
df1.info()

<class 'pandas.DataFrame'>
RangeIndex: 1542 entries, 0 to 1541
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   scope             1542 non-null   str  
 1   event_id          1542 non-null   str  
 2   event_name        1542 non-null   str  
 3   subject_category  1542 non-null   str  
 4   period            1542 non-null   str  
 5   event_date        1532 non-null   str  
 6   person_count      1542 non-null   int64
 7   related_event     605 non-null    str  
 8   detail_url        1542 non-null   str  
dtypes: int64(1), str(8)
memory usage: 108.6 KB


In [109]:
df1.head()

,scope,event_id,event_name,subject_category,period,event_date,person_count,related_event,detail_url
0,event_subject,ITKC_PH_1294A_0435,강동성전투,전쟁,고려,1218년(고종 5) 12월\r\n~ 1219년(고종 6) 1월,2,고려거란전쟁,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...
1,event_period,ITKC_PH_1294A_0435,강동성전투,전쟁,고려,1218년(고종 5) 12월\r\n~ 1219년(고종 6) 1월,2,고려거란전쟁,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...
2,event_period,ITKC_PH_1294A_0435,강동성전투,전쟁,고려,1218년(고종 5) 12월\r\n~ 1219년(고종 6) 1월,2,고려거란전쟁,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...
3,event_period,ITKC_PH_1294A_0436,거란의 항복,전쟁,고려,1219년(고종 6) 1월,2,고려거란전쟁,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...
4,event_period,ITKC_PH_1294A_0437,최충헌의 사망,정치인,고려,1219년(고종 6) 9월,4,최씨무인정권,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...


In [110]:
compare_cols = [col for col in df1.columns if col not in ["event_id", "scope"]]

diff_check = (
    df1
    .groupby("event_id")[compare_cols]
    .nunique(dropna=False)
)

diff_rows = diff_check[diff_check.gt(1).any(axis=1)]

diff_rows

,event_name,subject_category,period,event_date,person_count,related_event,detail_url
event_id,,,,,,,
ITKC_PH_1294A_0335,1,1,1,1,1,1,2
ITKC_PH_1294A_0336,1,1,1,1,1,1,2
ITKC_PH_1294A_0337,1,1,1,1,1,1,2
ITKC_PH_1294A_0338,1,1,1,1,1,1,2
ITKC_PH_1294A_0339,1,1,1,1,1,1,2
...,...,...,...,...,...,...,...
ITKC_PH_1305A_0330,1,1,1,1,1,1,2
ITKC_PH_1305A_0331,1,1,1,1,1,1,2
ITKC_PH_1305A_0332,1,1,1,1,1,1,2


In [81]:
df1 = df1.drop_duplicates(subset=["event_id"])

In [82]:
df1 = df1.drop(columns=['detail_url', 'person_count'])

In [83]:
df1.duplicated().sum()

np.int64(0)

In [84]:
df1.info()

<class 'pandas.DataFrame'>
Index: 600 entries, 0 to 1460
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   scope             600 non-null    str  
 1   event_id          600 non-null    str  
 2   event_name        600 non-null    str  
 3   subject_category  600 non-null    str  
 4   period            600 non-null    str  
 5   event_date        595 non-null    str  
 6   related_event     224 non-null    str  
dtypes: str(7)
memory usage: 37.5 KB


In [85]:
df1 = df1.drop(columns=['scope'])

In [86]:
df1.shape

(600, 6)

In [87]:
df1.head()

,event_id,event_name,subject_category,period,event_date,related_event
0,ITKC_PH_1294A_0435,강동성전투,전쟁,고려,1218년(고종 5) 12월\r\n~ 1219년(고종 6) 1월,고려거란전쟁
3,ITKC_PH_1294A_0436,거란의 항복,전쟁,고려,1219년(고종 6) 1월,고려거란전쟁
4,ITKC_PH_1294A_0437,최충헌의 사망,정치인,고려,1219년(고종 6) 9월,최씨무인정권
5,ITKC_PH_1294A_0438,의주적의 난,민란,고려,1219년(고종 6) 10월\r\n~ 1221년(고종 8) 3월,NaN
6,ITKC_PH_1294A_0439,이극인의 난,반란,고려,1224년(고종 11) 7월,최씨무인정권


In [88]:
df2.shape

(15392, 10)

In [89]:
df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 15392 entries, 0 to 15391
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   scope               15392 non-null  str    
 1   event_id            15392 non-null  str    
 2   event_name          15392 non-null  str    
 3   relation_type       15392 non-null  str    
 4   person_id           15392 non-null  str    
 5   person_name         15392 non-null  str    
 6   related_event_id    0 non-null      float64
 7   related_event_name  0 non-null      float64
 8   evidence_url        0 non-null      float64
 9   detail_url          15392 non-null  str    
dtypes: float64(3), str(7)
memory usage: 1.2 MB


In [90]:
df2.duplicated().sum()

np.int64(1556)

In [91]:
df2.head(5)

,scope,event_id,event_name,relation_type,person_id,person_name,related_event_id,related_event_name,evidence_url,detail_url
0,event_subject,ITKC_PH_1294A_0435,강동성전투,사건인물,P009535,김인경(金仁鏡),NaN,NaN,NaN,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...
1,event_subject,ITKC_PH_1294A_0435,강동성전투,사건인물,P058283,조충(趙冲),NaN,NaN,NaN,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...
2,event_period,ITKC_PH_1294A_0435,강동성전투,사건인물,P009535,김인경(金仁鏡),NaN,NaN,NaN,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...
3,event_period,ITKC_PH_1294A_0435,강동성전투,사건인물,P058283,조충(趙冲),NaN,NaN,NaN,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...
4,event_period,ITKC_PH_1294A_0435,강동성전투,사건인물,P009535,김인경(金仁鏡),NaN,NaN,NaN,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...


In [92]:
df2 = df2.drop_duplicates(subset=["event_id", "person_id"])

In [93]:
df2.info()

<class 'pandas.DataFrame'>
Index: 6918 entries, 0 to 14476
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   scope               6918 non-null   str    
 1   event_id            6918 non-null   str    
 2   event_name          6918 non-null   str    
 3   relation_type       6918 non-null   str    
 4   person_id           6918 non-null   str    
 5   person_name         6918 non-null   str    
 6   related_event_id    0 non-null      float64
 7   related_event_name  0 non-null      float64
 8   evidence_url        0 non-null      float64
 9   detail_url          6918 non-null   str    
dtypes: float64(3), str(7)
memory usage: 594.5 KB


In [94]:
df2.tail()

,scope,event_id,event_name,relation_type,person_id,person_name,related_event_id,related_event_name,evidence_url,detail_url
14472,event_subject,ITKC_PH_1305A_0334,심수경의 과거 동기,사건인물,P056339,조보(趙溥),NaN,NaN,NaN,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...
14473,event_subject,ITKC_PH_1305A_0334,심수경의 과거 동기,사건인물,P058193,조징(趙澄),NaN,NaN,NaN,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...
14474,event_subject,ITKC_PH_1305A_0334,심수경의 과거 동기,사건인물,P066169,홍천민(洪天民),NaN,NaN,NaN,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...
14475,event_subject,ITKC_PH_1305A_0334,심수경의 과거 동기,사건인물,P066576,황림(黃琳),NaN,NaN,NaN,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...
14476,event_subject,ITKC_PH_1305A_0334,심수경의 과거 동기,사건인물,P066959,황응규(黃應奎),NaN,NaN,NaN,https://db.itkc.or.kr/people/viewEvnt?gubun=ev...


In [95]:
df2 = df2.drop(columns=['scope', 'detail_url', 'evidence_url'])

In [96]:
df2.head()

,event_id,event_name,relation_type,person_id,person_name,related_event_id,related_event_name
0,ITKC_PH_1294A_0435,강동성전투,사건인물,P009535,김인경(金仁鏡),NaN,NaN
1,ITKC_PH_1294A_0435,강동성전투,사건인물,P058283,조충(趙冲),NaN,NaN
6,ITKC_PH_1294A_0436,거란의 항복,사건인물,P011067,김취려(金就礪),NaN,NaN
7,ITKC_PH_1294A_0436,거란의 항복,사건인물,P058283,조충(趙冲),NaN,NaN
8,ITKC_PH_1294A_0437,최충헌의 사망,사건인물,P008037,김약선(金若先),NaN,NaN


In [97]:
df2['related_event_id'].nunique()

0

In [98]:
df2['related_event_name'].nunique()

0

In [99]:
df2['relation_type'].unique()

<StringArray>
['사건인물']
Length: 1, dtype: str

In [100]:
df2 = df2.drop(columns=['related_event_id', 'related_event_name','relation_type'])

In [101]:
df2.head()

,event_id,event_name,person_id,person_name
0,ITKC_PH_1294A_0435,강동성전투,P009535,김인경(金仁鏡)
1,ITKC_PH_1294A_0435,강동성전투,P058283,조충(趙冲)
6,ITKC_PH_1294A_0436,거란의 항복,P011067,김취려(金就礪)
7,ITKC_PH_1294A_0436,거란의 항복,P058283,조충(趙冲)
8,ITKC_PH_1294A_0437,최충헌의 사망,P008037,김약선(金若先)


In [102]:
df1['subject_category'].unique()

<StringArray>
[               '전쟁',               '정치인',                '민란',
                '반란',    '반란,\r\n\r\n정치인',   '정치일반,\r\n\r\n국왕',
                '국왕',                '세자',              '인물기타',
     '반란,\r\n\r\n국왕',
 ...
                 '청',   '정치일반,\r\n\r\n붕당',  '옥사,\r\n\r\n고변/탄핵',
                '예송',     '예송,\r\n\r\n비빈',  '정치일반,\r\n\r\n정치인',
     '세자,\r\n\r\n비빈', '정치일반,\r\n\r\n묘정배향',     '옥사,\r\n\r\n세자',
              '사가독서']
Length: 119, dtype: str

In [103]:
df1['subject_category'].nunique()

119

## EDA 정리: event 데이터 중복과 `scope` 해석

현재 `df1`은 `itkc_events.csv`에서 읽은 사건 기본 정보이고, `df2`는 `itkc_event_relations.csv`에서 읽은 사건-인물 관계 정보다.

### 1. `scope` 해석

`df1.scope`는 최종 `Event` 노드의 역사 속성이라기보다, 같은 사건이 어떤 수집 경로에서 발견되었는지를 나타내는 값으로 본다.

```text
event_subject : 주제/분류 기준 수집 경로
event_period  : 시대 기준 수집 경로
```

현재 분포는 다음처럼 600개씩 동일하다.

```text
event_subject    600
event_period     600
```

즉 같은 사건이 `event_subject`, `event_period` 두 경로에서 중복 수집된 구조로 보인다.

### 2. `event_id` 기준 중복 제거 가능 여부 확인

`event_id`만 보고 바로 중복 제거하면 정보가 사라질 수 있으므로, 먼저 같은 `event_id` 안에서 `scope` 외의 값이 달라지는지 확인한다.

```python
compare_cols = [col for col in df1.columns if col not in ["event_id", "scope"]]

diff_check = (
    df1
    .groupby("event_id")[compare_cols]
    .nunique(dropna=False)
)

diff_rows = diff_check[diff_check.gt(1).any(axis=1)]
diff_rows
```

이 코드의 의미는 다음과 같다.

- `event_id`, `scope`를 제외한 컬럼만 비교한다.
- `event_id`별로 각 컬럼의 고유값 개수를 센다.
- 어떤 컬럼이라도 고유값 개수가 2개 이상이면 `diff_rows`에 남긴다.
- `diff_rows`가 비어 있으면 같은 `event_id` 안에서 다른 값은 `scope`뿐이라는 뜻이다.

현재 확인 결과는 각 비교 컬럼의 고유값 개수가 모두 `1`로 보인다. 따라서 `event_id` 기준으로 중복 제거해도 안전하다고 판단한다.

### 3. 최종 Event 노드 컬럼 판단

최종 `Event` 노드에는 `scope`를 넣지 않는다. `scope`는 수집 경로 메타데이터이고, 사건 자체의 속성은 아니기 때문이다.

1차 Event 노드에 남길 컬럼 후보는 다음과 같다.

| 컬럼 | 판단 | 이유 |
|---|---|---|
| `event_id` | 사용 | Event 고유 ID |
| `event_name` | 사용 | 사건명 |
| `subject_category` | 사용 | 사건 분류 후보, 이후 표준 카테고리 매핑 필요 |
| `period` | 사용 | 시대 텍스트, 이후 `Period` 연결 후보 |
| `event_date` | 사용 | 날짜 원문, 이후 연도/왕대 파싱 필요 |
| `related_event` | 사용 | `EventGroup` 후보 |
| `detail_url` | 선택 보존 | 출처 URL, 근거 추적에 유용 |
| `person_count` | 제외 가능 | `event_relation.csv`에서 다시 계산 가능 |
| `scope` | 제외 | 수집 경로 메타데이터 |

추천 정리 코드는 다음과 같다.

```python
event_df = (
    df1
    .drop_duplicates(subset=["event_id"])
    .drop(columns=["scope", "person_count"])
)
```

`detail_url`은 MVP에서 제거할 수도 있지만, 출처 추적을 위해 일단 남기는 편이 낫다.

### 4. `subject_category` 후속 처리

`subject_category`는 현재 고유값이 119개이고, 아래처럼 복수 분류가 문자열 안에 같이 들어간 경우가 있다.

```text
반란,\r\n\r\n정치인
정치일반,\r\n\r\n국왕
옥사,\r\n\r\n고변/탄핵
```

따라서 `event.csv.subject_category`는 바로 표준 카테고리로 쓰지 말고, 별도 `category_mapping.csv`에서 표준 `Category`로 매핑한다.

```text
event.csv.subject_category -> category_mapping.csv -> Category
```

즉, 이벤트 카테고리는 `history_terms.term_lk`와 직접 합치는 것이 아니라 공통 표준 카테고리 사전에 매핑하는 방식으로 처리한다.
